File: src/1_100/0086_Partition_List/Partition_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
    	if(head==null ||head.next==null) {
    		return head;
    	}
    	ListNode cur = head;
    	ListNode cur_min = new ListNode(-1);
    	ListNode pre_min = cur_min;
    	ListNode cur_max = new ListNode(-1);
    	ListNode pre_max = cur_max;
    	while(cur != null) {
    		if(cur.val < x) {
    			cur_min.next = cur;
    			cur_min = cur_min.next;
    		} else {
    			cur_max.next = cur;
    			cur_max = cur_max.next;
    		}
    		cur = cur.next;
    	}
    	cur_min.next = pre_max.next;
    	cur_max.next = null;
        return pre_min.next;
    }
}

File: src/1_100/0086_Partition_List/Partition_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        first,second = ListNode(-1),ListNode(-1)
        first.next,second.next,cur,a,b = head,head,head,first,second
        while cur:
            if cur.val<x:
                a.next,a,cur = cur,cur,cur.next
            else:
                b.next,b,cur = cur,cur,cur.next
        if a==first or b==second:
            return first.next if b==second else second.next
        a.next,b.next = second.next,None
        return first.next
        